{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6302107f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.extend(\n",
    "[\n",
    "    '/Users/fragoso/modules/',\n",
    "    '/Users/fragoso/modules/emeraldbgc',\n",
    "    '/Users/fragoso/modules/emeraldbgc/modules',\n",
    "]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "efdc6b69",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "import json\n",
    "import logging\n",
    "import os\n",
    "import re\n",
    "import pickle\n",
    "from itertools import groupby\n",
    "\n",
    "import numpy as np\n",
    "import tensorflow as tf\n",
    "from joblib import load\n",
    "\n",
    "# from emeraldbgc import __version__, _params\n",
    "_params = {\n",
    "    \"thBase\": 0.8733,\n",
    "    \"thBorder\": 0.98,\n",
    "    \"rmless\": 1,\n",
    "    \"fill\": 3,\n",
    "    \"greed\": {\"0\": 0.58, \"1\": 0.41, \"2\": 0.345, \"3\": 0},\n",
    "    \"shape\": 200,\n",
    "    \"ip_an\": [\"Pfam\", \"TIGRFAM\", \"PRINTS\", \"ProSitePatterns\", \"Gene3D\"]\n",
    "}\n",
    "\n",
    "log = logging.getLogger(f\"EMERALD.{__name__}\")\n",
    "\n",
    "\n",
    "class AnnotationFilesToEmerald:\n",
    "    \"\"\"Transform external tool's files into EMERALD STR\"\"\"\n",
    "\n",
    "    def __init__(self):\n",
    "\n",
    "        self.entriesDct = {}\n",
    "        self.contigsDct = {}\n",
    "        self.annDct = {}\n",
    "        self.typeDct = {}\n",
    "        self.annResults = {}\n",
    "        post_file = os.path.join(\n",
    "            os.path.dirname(os.path.abspath(__file__)),\n",
    "            \"..\",\n",
    "            \"models\",\n",
    "            \"post_filters.pickle\",\n",
    "        )\n",
    "        with open(post_file, \"rb\") as h:\n",
    "            self.vocab, self.typeModel, self.mbdoms = pickle.load(h)\n",
    "\n",
    "    def transformIPS(self, ipsFile):\n",
    "\n",
    "        if not os.path.isfile(ipsFile):\n",
    "            log.exception(f\"{ipsFile} file not found\")\n",
    "\n",
    "        with open(ipsFile, \"r\") as h:\n",
    "\n",
    "            lines = h.readlines()\n",
    "            fmt = \"gff\" if lines[0][:5] == \"##gff\" else \"tsv\"\n",
    "\n",
    "            for l in lines:\n",
    "        \n",
    "                if fmt == \"tsv\":\n",
    "\n",
    "                    spl = l.split(\"\\t\")\n",
    "                    self.entriesDct.setdefault(spl[0], []).append(\n",
    "                        spl[-2] if spl[-2] != '-' else spl[4]\n",
    "                    )\n",
    "                if fmt == \"gff\":\n",
    "                    spl = l.split(\"\\t\")\n",
    "                    if len(spl) > 3 and spl[2]=='protein_match':\n",
    "                        self.entriesDct.setdefault(spl[0],[]).append(\n",
    "                                re.split(\"InterPro:|\\\"\",spl[-1])[-2] if 'InterPro' in spl[-1]     else re.split(\"Name=|;\",spl[-1])[-2]\n",
    "                        )\n",
    "\n",
    "    def transformEmeraldHmm(self, hmmFile):\n",
    "\n",
    "        log.info(f\"processing {hmmFile}\")\n",
    "        if not os.path.isfile(hmmFile):\n",
    "            log.exception(f\"{hmmFile} file not found\")\n",
    "\n",
    "        with open(hmmFile, \"r\") as h:\n",
    "\n",
    "            for l in h:\n",
    "\n",
    "                if l[0] == \"#\":\n",
    "                    continue\n",
    "\n",
    "                spl = l.split()\n",
    "                self.entriesDct.setdefault(spl[3], []).append(spl[0])\n",
    "\n",
    "    def transformCDSpredToCDScontigs(self, cdsPredFile, f):\n",
    "\n",
    "        if not os.path.isfile(cdsPredFile):\n",
    "            log.exception(f\"{cdsPredFile} file not found\")\n",
    "\n",
    "        with open(cdsPredFile, \"r\") as h:\n",
    "\n",
    "            if f == \"fna\":\n",
    "\n",
    "                for l in h:\n",
    "\n",
    "                    if l[0] != \">\":\n",
    "                        continue\n",
    "\n",
    "                    spl = l.split()\n",
    "                    start, end = int(spl[2]), int(spl[4])\n",
    "\n",
    "                    self.contigsDct.setdefault(\n",
    "                        \"_\".join(spl[0].split(\"_\")[:-1])[1:], []\n",
    "                    ).append((spl[0][1:], (start, end)))\n",
    "\n",
    "            elif f == \"gbk\":\n",
    "\n",
    "                from Bio import SeqIO\n",
    "\n",
    "                recs = list(SeqIO.parse(open(cdsPredFile, \"r\"), \"gb\"))\n",
    "\n",
    "                for rec in recs:\n",
    "                    for f in rec.features:\n",
    "                        if f.type == \"CDS\":\n",
    "\n",
    "\n",
    "                            start, end = int(f.location.start) + 1, int(f.location.end)\n",
    "\n",
    "                            self.contigsDct.setdefault(rec.id, []).append(\n",
    "                                (\n",
    "                                    f.qualifiers[\"protein_id\"][0]\n",
    "                                    if \"protein_id\" in f.qualifiers\n",
    "                                    else f.qualifiers[\"locus_tag\"][0],\n",
    "                                    (start, end),\n",
    "                                )\n",
    "                            )\n",
    "\n",
    "    def buildMatrices(self):\n",
    "\n",
    "        for contig in self.contigsDct:\n",
    "\n",
    "            cdss = self.contigsDct[contig]\n",
    "\n",
    "            samps = (\n",
    "                len(cdss)\n",
    "                if not _params[\"shape\"]\n",
    "                else ((len(cdss) // _params[\"shape\"]) + 1) * _params[\"shape\"]\n",
    "            )\n",
    "\n",
    "            self.annDct[contig] = np.zeros((samps, len(self.vocab)))\n",
    "\n",
    "            for ix, cds in enumerate(cdss):\n",
    "\n",
    "                name = cds[0]\n",
    "                if name not in self.entriesDct:\n",
    "                    continue\n",
    "\n",
    "                modiAnn = [\n",
    "                    self.vocab[x] for x in self.entriesDct[name] if x in self.vocab\n",
    "                ]\n",
    "                self.annDct[contig][ix][modiAnn] = 1\n",
    "\n",
    "\n",
    "    def predictAnn(self, colapseFunc=max):\n",
    "\n",
    "        log.info(\"Predict BGC probability w/ TensorFlow\")\n",
    "        model_BGC_file = os.path.join(\n",
    "            os.path.dirname(os.path.abspath(__file__)), \"..\", \"models\", \"emerald.h5\"\n",
    "        )\n",
    "\n",
    "        model = tf.keras.models.load_model(\n",
    "            model_BGC_file, custom_objects={\"robustLoss\": {}}\n",
    "        )\n",
    "\n",
    "        for contig in self.annDct:\n",
    "            mat = self.annDct[contig]\n",
    "            xva_, vaIx = self.transformMat(mat)\n",
    "            vaS = (xva_.shape[0] // _params[\"shape\"]) * _params[\"shape\"]\n",
    "            xva1 = xva_[:vaS].reshape(vaS // _params[\"shape\"], _params[\"shape\"])\n",
    "            xva = np.append(\n",
    "                xva1,\n",
    "                [list(xva_[vaS:]) + [0] * (_params[\"shape\"] - (vaIx.shape[0] - vaS))],\n",
    "                axis=0,\n",
    "            )\n",
    "            self.predict_ = model.predict(xva)\n",
    "            predict = self.predict_.reshape(vaS + _params[\"shape\"])\n",
    "            predict_str_, nix = self.partialReStrMat(vaIx, predict, func=colapseFunc)\n",
    "            self.annResults[contig] = self.projectRes(predict_str_, nix, mat.shape[0])\n",
    "\n",
    "    def transformMat(self, mat):\n",
    "\n",
    "        scuash = np.array(list(map(lambda x: np.where(x == 1)[0], mat)))\n",
    "        nli, nix = [], []\n",
    "        for ix, x in enumerate(scuash):\n",
    "            if x.shape[0] == 0:\n",
    "                continue\n",
    "            nli.extend(x)\n",
    "            nix.extend([ix] * x.shape[0])\n",
    "        sups = [\n",
    "            (k, list(zip(*g))[1][0])\n",
    "            for k, g in groupby(zip(nli, nix), key=lambda x: x[0])\n",
    "        ]\n",
    "        if len(nli) == 0:\n",
    "            return np.array([0]), np.array([0])\n",
    "        a, b = list(zip(*sups))\n",
    "        a, b = np.array(a), np.array(b)\n",
    "        return a, b\n",
    "\n",
    "    def projectRes(self, res, ixes, lenOri):\n",
    "\n",
    "        nres_ = np.zeros(lenOri)\n",
    "        nres_[ixes] = res\n",
    "        prev = 0\n",
    "        rprev = 0\n",
    "        for ix, r in zip(ixes, res):\n",
    "            if ix - 1 == prev:\n",
    "                prev = ix\n",
    "                rprev = r\n",
    "                continue\n",
    "            for ixx, s in list(enumerate(np.linspace(rprev, r, ix - prev + 1)))[1:-1]:\n",
    "                nres_[prev + ixx] = s\n",
    "            prev = ix\n",
    "\n",
    "        return nres_\n",
    "\n",
    "    def partialReStrMat(self, Ix, res, func):\n",
    "\n",
    "        nres, nix = [], []\n",
    "        for k, g in groupby(zip(res, Ix), key=lambda x: x[1]):\n",
    "\n",
    "            gg = list(list(zip(*g))[0])\n",
    "            nix.append(k)\n",
    "            nres.append(func(gg))\n",
    "        return np.array(nres), np.array(nix)\n",
    "\n",
    "    def defineLooseClusters(self):\n",
    "\n",
    "        log.info(\"Define Clusters\")\n",
    "        self.bridged, self.looseClst, self.borderClst, self.typesClst = {}, {}, {}, {}\n",
    "\n",
    "        for contig in self.annResults:\n",
    "            self.looseClst[contig] = np.array(\n",
    "                self.rmLessThan(\n",
    "                    self.fillGap(\n",
    "                        np.where(self.annResults[contig] < _params[\"thBase\"], 0, 1),\n",
    "                        _params[\"fill\"],\n",
    "                    ),\n",
    "                    _params[\"rmless\"],\n",
    "                )\n",
    "            )\n",
    "            self.borderClst[contig] = np.where(\n",
    "                self.annResults[contig] < _params[\"thBorder\"], 0, 1\n",
    "            )\n",
    "            self.typesClst[contig] = np.empty(\n",
    "                len(self.annResults[contig]), dtype=object\n",
    "            )\n",
    "\n",
    "    def rmLessThan(self, contig, n):\n",
    "        newContig = []\n",
    "        for k, g in groupby(contig):\n",
    "            if k == 0:\n",
    "                newContig.extend(list(g))\n",
    "            else:\n",
    "                p = list(g)\n",
    "                if len(p) <= n:\n",
    "                    newContig.extend([0 for x in p])\n",
    "                else:\n",
    "                    newContig.extend(p)\n",
    "        return newContig\n",
    "\n",
    "    def fillGap(self, contig, gap):\n",
    "        newContig = []\n",
    "        for k, g in groupby(contig):\n",
    "            gg = list(g)\n",
    "            if k == 0 and len(gg) <= gap:\n",
    "                newContig.extend([1 for x in gg])\n",
    "            else:\n",
    "                newContig.extend(gg)\n",
    "        return newContig\n",
    "\n",
    "    def predictType(self, score=None, g=None):\n",
    "\n",
    "        log.info(\"Predict BGC classes\")\n",
    "\n",
    "        score = _params[\"greed\"][str(g)] if score == None else score\n",
    "        log.info(f\"Positive class model threshold: {score}\")\n",
    "        locations, matrix, typeLi = [], [], []\n",
    "        claa = [\n",
    "            \"Alkaloid\",\n",
    "            \"NRP\",\n",
    "            \"Polyketide\",\n",
    "            \"RiPP\",\n",
    "            \"Saccharide\",\n",
    "            \"Terpene\",\n",
    "            \"Other\",\n",
    "        ]\n",
    "\n",
    "        for contig in self.contigsDct:\n",
    "            for k, g in groupby(enumerate(self.looseClst[contig]), key=lambda x: x[1]):\n",
    "                if k == 0:\n",
    "                    continue\n",
    "                gg = list(list(zip(*g))[0])\n",
    "                tmat_ = np.sum(self.annDct[contig][gg], axis=0)\n",
    "                tmat = np.where(tmat_ > 0, 1, 0)\n",
    "                locations.append((contig, gg))\n",
    "                matrix.append(tmat)\n",
    "        if len(matrix) > 0:\n",
    "            pred = np.empty((len(matrix), len(claa)))\n",
    "            for nc, cla in enumerate(claa):\n",
    "                pred[:, nc] = self.typeModel[cla].predict_proba(matrix)[:, 1]\n",
    "        else:\n",
    "            pred = []\n",
    "\n",
    "        for ix, p in enumerate(pred):\n",
    "\n",
    "            contig, gg = locations[ix]\n",
    "\n",
    "            if not (np.max(p[[0, 1, 2, 3, 4, 5, 6]]) >= score):\n",
    "                self.looseClst[contig][gg] = 0\n",
    "                self.borderClst[contig][gg] = 0\n",
    "\n",
    "            else:\n",
    "                nearest_ = self.near_classifer(set(np.where(matrix[ix] == 1)[0]))\n",
    "                nearest = \"nearest_MiBIG={};nearest_MiBIG_class={};nearest_MiBIG_jaccardDistance={:.3f}\".format(\n",
    "                    nearest_[0], nearest_[1], nearest_[2]\n",
    "                )\n",
    "                self.typesClst[contig][gg] = nearest\n",
    "\n",
    "    def near_classifer(self, doms):\n",
    "        mb_ord = [(b, c, self.jacardDistance(doms, mbd)) for b, c, mbd in self.mbdoms]\n",
    "        nearest = sorted(mb_ord, key=lambda x: x[2])[0]\n",
    "        return nearest\n",
    "\n",
    "    def jacardDistance(self, a, b):\n",
    "        try:\n",
    "            return 1 - ((len(set(a) & set(b))) / len(set(a) | set(b)))\n",
    "        except:\n",
    "            return 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a74d2c0f",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "import logging\n",
    "import os\n",
    "import sys\n",
    "import subprocess\n",
    "\n",
    "# from emeraldbgc import _params\n",
    "# log = logging.getLogger(f\"EMERALD.{__name__}\")\n",
    "\n",
    "from distutils.spawn import find_executable\n",
    "\n",
    "\n",
    "class Preprocess:\n",
    "    \"\"\"External tools needed for emerald bgc detection\"\"\"\n",
    "\n",
    "    def __init__(self, seq_file, ip_file, meta, cpus, outdir):\n",
    "\n",
    "        self.seq_file = seq_file\n",
    "        self.ip_file = ip_file\n",
    "        self.meta = meta\n",
    "        self.cpus = int(cpus)\n",
    "        self.outdir = outdir if outdir else \"temp\"\n",
    "\n",
    "    def runProdigal(self):\n",
    "        \"\"\"Predict genes using prodigal\"\"\"\n",
    "        log.info(\"Progial gene prediction...\")\n",
    "\n",
    "        if not find_executable(\"prodigal\"):\n",
    "            log.exception(\"Parodigal is not installed or in PATH\")\n",
    "\n",
    "        if not os.path.isfile(self.seq_file):\n",
    "            log.exception(f\"{self.seq_file} file not found\")\n",
    "\n",
    "        outFaa = os.path.join(\n",
    "            self.outdir, \"{}.prodigal.faa\".format(os.path.basename(self.seq_file))\n",
    "        )\n",
    "        cmd = [\"prodigal\", \"-i\", self.seq_file, \"-a\", outFaa] + (\n",
    "            [\"-p\", \"meta\"] if self.meta == \"True\" else []\n",
    "        )\n",
    "\n",
    "        outs, errs = subprocess.Popen(\n",
    "            cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE\n",
    "        ).communicate()\n",
    "\n",
    "        if \"Error:  Sequence must be 20000 characters\" in errs.decode(\"utf8\"):\n",
    "\n",
    "            log.info(\n",
    "                \"Sequence must be 20000 characters when running Prodigal in normal mode. Trying -p meta\"\n",
    "            )\n",
    "\n",
    "            cmd = [\"prodigal\", \"-i\", self.seq_file, \"-a\", outFaa, \"-m\"] + ([\"-p\", \"meta\"])\n",
    "\n",
    "            outs, errs = subprocess.Popen(\n",
    "                cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE\n",
    "            ).communicate()\n",
    "\n",
    "        log.info(errs.decode(\"utf8\"))\n",
    "\n",
    "        \"\"\" Remove asterixs from faa\n",
    "            \"\"\"\n",
    "        log.info(\"Removing asterix from prodigal faa\")\n",
    "        with open(outFaa, \"r\") as h:\n",
    "            noAstFaa = [x.replace(\"*\", \"\") for x in h]\n",
    "        with open(outFaa, \"w\") as h:\n",
    "            for l in noAstFaa:\n",
    "                h.write(f\"{l}\")\n",
    "\n",
    "        return os.path.abspath(outFaa)\n",
    "\n",
    "    def gbkToProdigal(self):\n",
    "        \"\"\"Transform gbk to faa. This enables preprocessing with sequences\"\"\"\n",
    "        log.info(\"write gbk as faa\")\n",
    "        from Bio import SeqIO\n",
    "\n",
    "        recs = list(SeqIO.parse(open(self.seq_file, \"r\"), \"gb\"))\n",
    "\n",
    "        base = os.path.basename(self.seq_file)\n",
    "\n",
    "        outFaa = os.path.join(\n",
    "            self.outdir, \"{}.prodigal.faa\".format(os.path.basename(self.seq_file))\n",
    "        )\n",
    "\n",
    "        with open(outFaa, \"w\") as h:\n",
    "\n",
    "            for rec in recs:\n",
    "                ct = 0\n",
    "                for f in rec.features:\n",
    "\n",
    "                    \n",
    "                    if f.type == \"CDS\" and \"translation\" in f.qualifiers:\n",
    "                        ct += 1\n",
    "                        pid = (\n",
    "                            f.qualifiers[\"protein_id\"][0]\n",
    "                            if \"protein_id\" in f.qualifiers\n",
    "                            else f.qualifiers[\"locus_tag\"][0]\n",
    "                        )\n",
    "                        seq = f.qualifiers[\"translation\"][0]\n",
    "                        h.write(f\">{pid}\\n{seq}\\n\")\n",
    "                        \n",
    "                if ct == 0:\n",
    "                    log.info(\"{} CDS found with translation in {}\".format(ct, rec.name) )\n",
    "\n",
    "        return os.path.abspath(outFaa)\n",
    "\n",
    "    def check_fmt(self):\n",
    "        \"\"\" Evaluate if input format is FNA or GBK\"\"\"\n",
    "        with open(self.seq_file) as h:\n",
    "            if h.readline()[0] == \">\":\n",
    "                self.fmt = \"fna\"\n",
    "                log.info(\"FASTA sequence file detected\")\n",
    "            else:\n",
    "                self.fmt = \"gbk\"\n",
    "                log.info(f\"FASTA sequence file NOT detected; trying GBK\")\n",
    "\n",
    "    def process_sequence(self):\n",
    "        \"\"\" CDS prediction on sequence file\"\"\"\n",
    "        if 'linux' not in sys.platform and self.ip_file == None:\n",
    "            log.info(\"internal run of InterProScan only available for Linux OS. Make sure to use --ip-file option\")\n",
    "            raise ValueError('Non Linux OS must be run with --ip-file option')\n",
    "\n",
    "        self.check_fmt()\n",
    "        if self.fmt == \"fna\":\n",
    "            self.outFaa = self.runProdigal()\n",
    "        elif self.fmt == \"gbk\":\n",
    "            self.outFaa = self.gbkToProdigal()\n",
    "        else:\n",
    "            log.info(\"missing sequence file format\")\n",
    "\n",
    "        ih_f = self.runHmmScan()\n",
    "        ip_f = self.runInterproscan() if self.ip_file == None else self.ip_file\n",
    "        return self.outFaa, ip_f, ih_f\n",
    "\n",
    "    def runHmmScan(self):\n",
    "        \"\"\"annotate functionally with emerald hmm library and hmmScan\"\"\"\n",
    "        log.info(\"emerald functional annotation...\")\n",
    "        hmmLib = os.path.join(\n",
    "            os.path.dirname(os.path.abspath(__file__)),\n",
    "            \"..\",\n",
    "            \"models\",\n",
    "            \"hmm_lib\",\n",
    "            \"emerald.hmm\",\n",
    "        )\n",
    "\n",
    "        if not find_executable(\"hmmscan\"):\n",
    "            log.exception(\"hmmscan is not installed or in PATH\")\n",
    "\n",
    "        if not os.path.isfile(self.outFaa):\n",
    "            log.exception(f\"{self.outFaa} file not found\")\n",
    "\n",
    "        if not os.path.isfile(hmmLib):\n",
    "            log.exception(f\"{hmmLib} file not found\")\n",
    "\n",
    "        outTsv = os.path.join(\n",
    "            self.outdir, \"{}.emerald.tsv\".format(os.path.basename(self.outFaa))\n",
    "        )\n",
    "\n",
    "        cmd = (\n",
    "            [\"hmmscan\", \"--domtblout\", outTsv, \"--cut_ga\"]\n",
    "            + ([\"--cpu\", str(self.cpus)] if self.cpus else [])\n",
    "            + ([hmmLib, self.outFaa])\n",
    "        )\n",
    "        log.info(\" \".join(cmd))\n",
    "        try:\n",
    "            outs, errs = subprocess.Popen(\n",
    "                cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE\n",
    "            ).communicate()\n",
    "            log.info(errs.decode(\"utf8\"))\n",
    "        except subprocess.CalledProcessError as err:\n",
    "            log.exception(err.output)\n",
    "\n",
    "        return os.path.abspath(outTsv)\n",
    "\n",
    "    def runInterproscan(self):\n",
    "        \"\"\"annotate functionally with InterproScan\"\"\"\n",
    "        log.info(\"InterProScan\")\n",
    "\n",
    "        if not find_executable('interproscan.sh'):\n",
    "            log.exception(f\"interproscan.sh not found, only available for Linux OS\")\n",
    "\n",
    "        if not os.path.isfile(self.outFaa):\n",
    "            log.exception(f\"self.outFaa file not found\")\n",
    "\n",
    "        outGff = os.path.join(\n",
    "            self.outdir, \"{}.ip.tsv\".format(os.path.basename(self.outFaa))\n",
    "        )\n",
    "        cmd = (\n",
    "            [\"interproscan.sh\", \"-i\", self.outFaa, \"-o\", outGff, \"-f\", \"TSV\"]\n",
    "            + ([\"-appl\", \",\".join(_params[\"ip_an\"])] if _params[\"ip_an\"] else [])\n",
    "            + ([\"-cpu\", str(self.cpus)] if self.cpus else [])\n",
    "        )\n",
    "        log.info(\" \".join(cmd))\n",
    "\n",
    "        outs, errs = subprocess.Popen(\n",
    "            cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE\n",
    "        ).communicate()\n",
    "        log.info(outs.decode(\"utf8\"))\n",
    "        log.info(errs.decode(\"utf8\"))\n",
    "\n",
    "        return os.path.abspath(outGff)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c7e669f8",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "import json\n",
    "import logging\n",
    "import os\n",
    "import pickle\n",
    "from itertools import groupby\n",
    "\n",
    "import numpy as np\n",
    "import tensorflow as tf\n",
    "from joblib import load\n",
    "\n",
    "from emeraldbgc import __version__\n",
    "\n",
    "log = logging.getLogger(f\"EMERALD.{__name__}\")\n",
    "\n",
    "\n",
    "class Outputs:\n",
    "\n",
    "    \"\"\" write results in file \"\"\"\n",
    "\n",
    "    def __init__(self, obj, minimal_output, refined_borders, outfile):\n",
    "\n",
    "        self.annotation = obj\n",
    "        self.minimal = minimal_output\n",
    "        self.ref_b = refined_borders\n",
    "        self.outfile = outfile\n",
    "        self.gff3 = []\n",
    "        self.clusts = []\n",
    "        self.writeGff3()\n",
    "\n",
    "    def writeGff3(self):\n",
    "\n",
    "        log.info(\"build GFF3\")\n",
    "        log.info(self.outfile)\n",
    "        type_code = {\n",
    "            0: \"Alkaloid\",\n",
    "            1: \"NRP\",\n",
    "            2: \"Polyketide\",\n",
    "            3: \"RiPP\",\n",
    "            4: \"Saccharide\",\n",
    "            5: \"Terpene\",\n",
    "            6: \"Other\",\n",
    "        }\n",
    "\n",
    "        for contig in self.annotation.looseClst:\n",
    "\n",
    "            if self.minimal != \"True\":\n",
    "                for ix, f in enumerate(self.annotation.contigsDct[contig]):\n",
    "\n",
    "                    ID, (start, end) = f\n",
    "                    emrldProb = \"{:.3f}\".format(self.annotation.annResults[contig][ix])\n",
    "                    self.gff3.append(\n",
    "                        f\"{contig}\\tEMERALDv{__version__}\\tCDS\\t{start}\\t{end}\\t.\\t.\\t.\\tID={ID};emerald_probability={emrldProb}\"\n",
    "                    )\n",
    "\n",
    "            ct = 1\n",
    "            for k, g in groupby(\n",
    "                enumerate(\n",
    "                    self.annotation.looseClst[contig][\n",
    "                        : len(self.annotation.contigsDct[contig])\n",
    "                    ]\n",
    "                ),\n",
    "                key=lambda x: x[1],\n",
    "            ):\n",
    "\n",
    "                if k == 0:\n",
    "                    continue\n",
    "\n",
    "                ID = f\"{contig}_emrld_{ct}\"\n",
    "                ct += 1\n",
    "\n",
    "                gg = list(list(zip(*g))[0])\n",
    "\n",
    "                start, end = (\n",
    "                    self.annotation.contigsDct[contig][gg[0]][1][0],\n",
    "                    self.annotation.contigsDct[contig][gg[-1]][1][1],\n",
    "                )\n",
    "\n",
    "                edge = \"{}{}\".format(\n",
    "                    1 if gg[0] == 0 else 0,\n",
    "                    1 if gg[-1] == len(self.annotation.contigsDct[contig]) - 1 else 0,\n",
    "                )\n",
    "\n",
    "                typs = self.annotation.typesClst[contig][gg[0]]\n",
    "                # typs = \"nearest_MiBIG={};nearest_MiBIG_class={};nearest_MiBIG_jacardDistance={}\".format(typs[0],typs[1],typs[2])\n",
    "                self.gff3.append(\n",
    "                    f\"{contig}\\tEMERALDv{__version__}\\tCLUSTER\\t{start}\\t{end}\\t.\\t.\\t.\\tID={ID};{typs};partial={edge}\"\n",
    "                )\n",
    "\n",
    "                if self.ref_b == \"True\":\n",
    "\n",
    "                    ct2 = 1\n",
    "                    for k2, g2 in groupby(\n",
    "                        zip(gg, self.annotation.borderClst[contig][gg]),\n",
    "                        key=lambda x: x[1],\n",
    "                    ):\n",
    "                        if k2 == 0:\n",
    "                            continue\n",
    "                        gg2 = list(list(zip(*g2))[0])\n",
    "                        start, end = (\n",
    "                            self.annotation.contigsDct[contig][gg2[0]][1][0],\n",
    "                            self.annotation.contigsDct[contig][gg2[-1]][1][1],\n",
    "                        )\n",
    "                        edge = \"{}{}\".format(\n",
    "                            1 if gg[0] == 0 else 0,\n",
    "                            1\n",
    "                            if gg[-1] == len(self.annotation.contigsDct[contig]) - 1\n",
    "                            else 0,\n",
    "                        )\n",
    "\n",
    "                        self.gff3.append(\n",
    "                            f\"{contig}\\tEMERALDv{__version__}\\tCLUSTER_border\\t{start}\\t{end}\\t.\\t.\\t.\\tID={ID}_{ct2};{typs};partial={edge}\"\n",
    "                        )\n",
    "                        ct2 += 1\n",
    "\n",
    "        log.info(f\"Writing output to file {self.outfile}\")\n",
    "        with open(self.outfile, \"w\") as h:\n",
    "\n",
    "            h.write(\"##gff-version 3\\n\")\n",
    "\n",
    "            key = lambda x: (\n",
    "                x.split(\"\\t\")[0],\n",
    "                int(x.split(\"\\t\")[3]),\n",
    "                \"Z\" if x.split(\"\\t\")[2] == \"CDS\" else x.split(\"\\t\")[2],\n",
    "            )\n",
    "            for l in sorted(self.gff3, key=key):\n",
    "\n",
    "                h.write(f\"{l}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "95991321",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/Users/fragoso/modules/emeraldbgc/emeraldbgc/modules',\n",
       " '/Users/fragoso/miniconda3/lib/python38.zip',\n",
       " '/Users/fragoso/miniconda3/lib/python3.8',\n",
       " '/Users/fragoso/miniconda3/lib/python3.8/lib-dynload',\n",
       " '',\n",
       " '/Users/fragoso/miniconda3/lib/python3.8/site-packages',\n",
       " '/Users/fragoso/miniconda3/lib/python3.8/site-packages/IPython/extensions',\n",
       " '/Users/fragoso/.ipython',\n",
       " '/Users/fragoso/modules/',\n",
       " '/Users/fragoso/modules/emeraldbgc',\n",
       " '/Users/fragoso/modules/emeraldbgc/modules']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sys.path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c5b06a89",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/fragoso/modules/emeraldbgc\n"
     ]
    }
   ],
   "source": [
    "cd /Users/fragoso/modules/emeraldbgc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7ef2679d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import argparse\n",
    "import glob\n",
    "import logging\n",
    "import os\n",
    "import sys\n",
    "\n",
    "# from emeraldbgc import __version__\n",
    "\n",
    "# sys.path.append(os.path.dirname(os.path.abspath('.')))\n",
    "\n",
    "# from modules.BGCdetection import AnnotationFilesToEmerald\n",
    "# from modules.Preproc import Preprocess\n",
    "# from modules.WriteOutput import Outputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a1e1fa67",
   "metadata": {},
   "outputs": [],
   "source": [
    "args = [\n",
    "    '--ip-file',\n",
    "    '/Users/fragoso/Downloads/ws/bench/files/comp/CM000950.fna.prodigal.faa.ip.tsv',\n",
    "    '/Users/fragoso/Downloads/ws/bench/files/comp/CM000950.gb'\n",
    "]\n",
    "__file__ = '/Users/fragoso/modules/emeraldbgc/emeraldbgc/modules/dax.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0ca9f178",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOG_FILE: /Users/fragoso/modules/emeraldbgc/CM000950.gb.emerald/emerald.log\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-2-935855b32f1a>:183: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray\n",
      "  scuash = np.array(list(map(lambda x: np.where(x == 1)[0], mat)))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EMERALD succesful\n"
     ]
    }
   ],
   "source": [
    "# def main(args=None):\n",
    "\n",
    "parser = argparse.ArgumentParser(description=\"EMERALD. SMBGC detection tool\")\n",
    "parser.add_argument(\n",
    "    \"seq_file\",\n",
    "    type=str,\n",
    "    help=\"input nucleotide sequence file. FASTA or GBK. mandatory\",\n",
    "    metavar=\"SEQUENCE_FILE\",\n",
    ")\n",
    "parser.add_argument(\n",
    "    \"-v\",\n",
    "    \"--version\",\n",
    "    action=\"version\",\n",
    "    help=\"Show the version number and exit.\",\n",
    "    version=f\"EMERALD {__version__}\",\n",
    ")\n",
    "parser.add_argument(\n",
    "    \"--ip-file\",\n",
    "    dest=\"ip_file\",\n",
    "    default=None,\n",
    "    type=str,\n",
    "    help=\"Optional, preprocessed InterProScan GFF3 output file. Requires a GBK file as SEQUENCE_FILE. The GBK must have CDS as features, and \\\"protein_id\\\" matching the ids in the InterProScan file\",\n",
    "    metavar=\"FILE\",\n",
    ")\n",
    "parser.add_argument(\n",
    "    \"--greed\",\n",
    "    dest=\"greed\",\n",
    "    default=2,\n",
    "    type=int,\n",
    "    help=\"Level of greediness. 0,1,2,3 [default 2]\",\n",
    "    metavar=\"INT\",\n",
    ")\n",
    "parser.add_argument(\n",
    "    \"--score\",\n",
    "    dest=\"score\",\n",
    "    default=None,\n",
    "    type=float,\n",
    "    help=\"validation filter threshold. overrides --greed\",\n",
    "    metavar=\"FLOAT\",\n",
    ")\n",
    "parser.add_argument(\n",
    "    \"--meta\",\n",
    "    dest=\"meta\",\n",
    "    default=\"True\",\n",
    "    type=str,\n",
    "    help=\"prodigal option meta [default True]\",\n",
    "    metavar=\"True|False\",\n",
    ")\n",
    "parser.add_argument(\n",
    "    \"--outdir\",\n",
    "    default=os.getcwd(),\n",
    "    dest=\"outdir\",\n",
    "    type=str,\n",
    "    help=\"output directory [default $PWD/SEQUENCE_FILE.emerald]\",\n",
    "    metavar=\"DIRECTORY\",\n",
    ")\n",
    "parser.add_argument(\n",
    "    \"--outfile\",\n",
    "    dest=\"outfile\",\n",
    "    type=str,\n",
    "    help=\"output file [default outdir/SEQUENCE_FILE.emerald.gff]\",\n",
    "    metavar=\"FILE\",\n",
    ")\n",
    "parser.add_argument(\n",
    "    \"--minimal\",\n",
    "    dest=\"minimal_out\",\n",
    "    default=\"True\",\n",
    "    type=str,\n",
    "    help=\"minimal output in a gff3 file [default True]\",\n",
    "    metavar=\"True|False\",\n",
    ")\n",
    "parser.add_argument(\n",
    "    \"--refined\",\n",
    "    dest=\"ref_b\",\n",
    "    default=\"False\",\n",
    "    type=str,\n",
    "    help=\"annotate high probability borders [default False]\",\n",
    "    metavar=\"True|False\",\n",
    ")\n",
    "parser.add_argument(\n",
    "    \"--cpu\",\n",
    "    dest=\"cpu\",\n",
    "    default=1,\n",
    "    type=int,\n",
    "    help=\"cpus for INTERPROSCAN and HMMSCAN\",\n",
    "    metavar=\"INT\",\n",
    ")\n",
    "\n",
    "args = parser.parse_args(args)\n",
    "\n",
    "basef = args.seq_file\n",
    "base = os.path.basename(basef)\n",
    "\n",
    "outdir = os.path.join(args.outdir, f\"{base}.emerald\")\n",
    "\n",
    "os.makedirs(outdir, exist_ok=True)\n",
    "\n",
    "logging.basicConfig(\n",
    "    filename=f\"{outdir}/emerald.log\",\n",
    "    level=logging.DEBUG,\n",
    "    format=\"%(asctime)s %(levelname)s %(name)s %(message)s\",\n",
    ")\n",
    "print(f\"LOG_FILE: {outdir}/emerald.log\")\n",
    "log = logging.getLogger(\"EMERALD\")\n",
    "log.info(\n",
    "    f\"\"\"\n",
    "******\n",
    "EMERALD v{__version__}\n",
    "******\"\"\"\n",
    ")\n",
    "log.info(f\"outdir: {outdir}\")\n",
    "\n",
    "log.info(\"preprocessing files\")\n",
    "preprocess = Preprocess(\n",
    "        os.path.abspath(args.seq_file), \n",
    "        args.ip_file, \n",
    "        args.meta, \n",
    "        args.cpu, \n",
    "        outdir\n",
    ")\n",
    "\n",
    "\n",
    "prodigal_file, ips_file, hmm_file = preprocess.process_sequence()\n",
    "\n",
    "log.info(\"EMERALD process\")\n",
    "annotate = AnnotationFilesToEmerald()\n",
    "\n",
    "log.info(\"transform interpro file\")\n",
    "annotate.transformIPS(ips_file)\n",
    "\n",
    "log.info(\"transform inhouse hmm file\")\n",
    "annotate.transformEmeraldHmm(hmm_file)\n",
    "\n",
    "log.info(\"transform proteins file\")\n",
    "annotate.transformCDSpredToCDScontigs(\n",
    "        prodigal_file if preprocess.fmt == \"fna\" else args.seq_file,\n",
    "        preprocess.fmt)\n",
    "\n",
    "log.info(\"transform dicts to np matrices\")\n",
    "annotate.buildMatrices()\n",
    "\n",
    "log.info(\"predict bgc regions\")\n",
    "annotate.predictAnn()\n",
    "\n",
    "log.info(\"define clusters\")\n",
    "annotate.defineLooseClusters()\n",
    "\n",
    "log.info(\"post-processing filters and type classification\")\n",
    "annotate.predictType(score=args.score, g=args.greed)\n",
    "\n",
    "log.info(\"write output file file\")\n",
    "outp = Outputs(\n",
    "    annotate,\n",
    "    args.minimal_out,\n",
    "    args.ref_b,\n",
    "    args.outfile if args.outfile else f\"{outdir}/{base}.emerald.full.gff\",\n",
    ")\n",
    "\n",
    "log.info(\"EMERALD succesful\")\n",
    "print(\"EMERALD succesful\")\n",
    "\n",
    "\n",
    "# if __name__ == \"__main__\":\n",
    "#     main(sys.argv[1:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "16345f3e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/fragoso/modules/emeraldbgc/CM000950.gb.emerald/CM000950.gb.emerald.full.gff'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f\"{outdir}/{base}.emerald.full.gff\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "3b04eb2c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CM000950\tEMERALDv0+unknown\tCLUSTER\t235657\t280316\t.\t.\t.\tID=CM000950_emrld_1;nearest_MiBIG=BGC0000804;nearest_MiBIG_class=Saccharide;nearest_MiBIG_jaccardDistance=0.810;partial=00\n",
      "CM000950\tEMERALDv0+unknown\tCLUSTER\t651667\t716832\t.\t.\t.\tID=CM000950_emrld_2;nearest_MiBIG=BGC0001767;nearest_MiBIG_class=NRP;nearest_MiBIG_jaccardDistance=0.564;partial=00\n",
      "CM000950\tEMERALDv0+unknown\tCLUSTER\t747262\t777103\t.\t.\t.\tID=CM000950_emrld_3;nearest_MiBIG=BGC0000663;nearest_MiBIG_class=Terpene;nearest_MiBIG_jaccardDistance=0.678;partial=00\n",
      "CM000950\tEMERALDv0+unknown\tCLUSTER\t917061\t974020\t.\t.\t.\tID=CM000950_emrld_4;nearest_MiBIG=BGC0001496;nearest_MiBIG_class=RiPP;nearest_MiBIG_jaccardDistance=0.833;partial=00\n",
      "CM000950\tEMERALDv0+unknown\tCLUSTER\t1451219\t1472608\t.\t.\t.\tID=CM000950_emrld_5;nearest_MiBIG=BGC0001496;nearest_MiBIG_class=RiPP;nearest_MiBIG_jaccardDistance=0.806;partial=00\n",
      "CM000950\tEMERALDv0+unknown\tCLUSTER\t1783748\t1853946\t.\t.\t.\tID=CM000950_emrld_6;nearest_MiBIG=BGC0001062;nearest_MiBIG_class=Polyketide;nearest_MiBIG_jaccardDistance=0.735;partial=00\n",
      "CM000950\tEMERALDv0+unknown\tCLUSTER\t1894864\t1953271\t.\t.\t.\tID=CM000950_emrld_7;nearest_MiBIG=BGC0001179;nearest_MiBIG_class=Polyketide;nearest_MiBIG_jaccardDistance=0.829;partial=00\n",
      "CM000950\tEMERALDv0+unknown\tCLUSTER\t2037759\t2065499\t.\t.\t.\tID=CM000950_emrld_8;nearest_MiBIG=BGC0001572;nearest_MiBIG_class=Other;nearest_MiBIG_jaccardDistance=0.859;partial=00\n",
      "CM000950\tEMERALDv0+unknown\tCLUSTER\t2423846\t2434683\t.\t.\t.\tID=CM000950_emrld_9;nearest_MiBIG=BGC0000496;nearest_MiBIG_class=RiPP;nearest_MiBIG_jaccardDistance=0.591;partial=00\n",
      "CM000950\tEMERALDv0+unknown\tCLUSTER\t2974377\t2994307\t.\t.\t.\tID=CM000950_emrld_10;nearest_MiBIG=BGC0000908;nearest_MiBIG_class=Other;nearest_MiBIG_jaccardDistance=0.881;partial=00\n",
      "CM000950\tEMERALDv0+unknown\tCLUSTER\t3638063\t3681381\t.\t.\t.\tID=CM000950_emrld_11;nearest_MiBIG=BGC0001537;nearest_MiBIG_class=Polyketide;nearest_MiBIG_jaccardDistance=0.688;partial=00\n",
      "CM000950\tEMERALDv0+unknown\tCLUSTER\t3823745\t3855691\t.\t.\t.\tID=CM000950_emrld_12;nearest_MiBIG=BGC0001912;nearest_MiBIG_class=NRP;nearest_MiBIG_jaccardDistance=0.757;partial=00\n",
      "CM000950\tEMERALDv0+unknown\tCLUSTER\t3969841\t3986931\t.\t.\t.\tID=CM000950_emrld_13;nearest_MiBIG=BGC0000501;nearest_MiBIG_class=RiPP;nearest_MiBIG_jaccardDistance=0.741;partial=00\n",
      "CM000950\tEMERALDv0+unknown\tCLUSTER\t4525353\t4579641\t.\t.\t.\tID=CM000950_emrld_14;nearest_MiBIG=BGC0001368;nearest_MiBIG_class=NRP;nearest_MiBIG_jaccardDistance=0.770;partial=00\n",
      "CM000950\tEMERALDv0+unknown\tCLUSTER\t4754043\t4780527\t.\t.\t.\tID=CM000950_emrld_15;nearest_MiBIG=BGC0001483;nearest_MiBIG_class=Other;nearest_MiBIG_jaccardDistance=0.541;partial=00\n",
      "CM000950\tEMERALDv0+unknown\tCLUSTER\t5129720\t5146039\t.\t.\t.\tID=CM000950_emrld_16;nearest_MiBIG=BGC0000940;nearest_MiBIG_class=Other;nearest_MiBIG_jaccardDistance=0.630;partial=00\n",
      "CM000950\tEMERALDv0+unknown\tCLUSTER\t5271093\t5273041\t.\t.\t.\tID=CM000950_emrld_17;nearest_MiBIG=BGC0000683;nearest_MiBIG_class=Terpene;nearest_MiBIG_jaccardDistance=0.500;partial=00\n",
      "CM000950\tEMERALDv0+unknown\tCLUSTER\t5341020\t5364603\t.\t.\t.\tID=CM000950_emrld_18;nearest_MiBIG=BGC0000025;nearest_MiBIG_class=Polyketide;nearest_MiBIG_jaccardDistance=0.750;partial=00\n",
      "CM000950\tEMERALDv0+unknown\tCLUSTER\t5464903\t5476144\t.\t.\t.\tID=CM000950_emrld_19;nearest_MiBIG=BGC0000581;nearest_MiBIG_class=RiPP;nearest_MiBIG_jaccardDistance=0.765;partial=00\n",
      "CM000950\tEMERALDv0+unknown\tCLUSTER\t5516963\t5527531\t.\t.\t.\tID=CM000950_emrld_20;nearest_MiBIG=BGC0000284;nearest_MiBIG_class=Polyketide;nearest_MiBIG_jaccardDistance=0.765;partial=00\n",
      "CM000950\tEMERALDv0+unknown\tCLUSTER\t6164083\t6192894\t.\t.\t.\tID=CM000950_emrld_21;nearest_MiBIG=BGC0001942;nearest_MiBIG_class=NRP Polyketide;nearest_MiBIG_jaccardDistance=0.768;partial=00\n",
      "CM000950\tEMERALDv0+unknown\tCLUSTER\t6537370\t6557814\t.\t.\t.\tID=CM000950_emrld_22;nearest_MiBIG=BGC0000431;nearest_MiBIG_class=NRP;nearest_MiBIG_jaccardDistance=0.711;partial=00\n",
      "CM000950\tEMERALDv0+unknown\tCLUSTER\t6792113\t6832578\t.\t.\t.\tID=CM000950_emrld_23;nearest_MiBIG=BGC0001817;nearest_MiBIG_class=Other;nearest_MiBIG_jaccardDistance=0.784;partial=00\n",
      "CM000950\tEMERALDv0+unknown\tCLUSTER\t7526991\t7555965\t.\t.\t.\tID=CM000950_emrld_24;nearest_MiBIG=BGC0000842;nearest_MiBIG_class=Other;nearest_MiBIG_jaccardDistance=0.850;partial=00\n",
      "CM000950\tEMERALDv0+unknown\tCLUSTER\t7693037\t7865528\t.\t.\t.\tID=CM000950_emrld_25;nearest_MiBIG=BGC0001073;nearest_MiBIG_class=NRP Polyketide;nearest_MiBIG_jaccardDistance=0.648;partial=00\n",
      "CM000950\tEMERALDv0+unknown\tCLUSTER\t7873080\t8025920\t.\t.\t.\tID=CM000950_emrld_26;nearest_MiBIG=BGC0000238;nearest_MiBIG_class=Polyketide;nearest_MiBIG_jaccardDistance=0.671;partial=00\n"
     ]
    }
   ],
   "source": [
    "!grep \"TER\\s\" /Users/fragoso/modules/emeraldbgc/CM000950.gb.emerald/CM000950.gb.emerald.full.gff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "94340a82",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CM000950\tEMERALDv0.1\tCLUSTER\t235657\t280316\t.\t.\t.\tID=CM000950_emrld_1;Class_Probability=Alkaloid:0.020,NRP:0.080,Polyketide:0.400,RiPP:0.070,Saccharide:0.120,Terpene:0.210,Other:0.240;Edge=False\n",
      "CM000950\tEMERALDv0.1\tCLUSTER\t651667\t716832\t.\t.\t.\tID=CM000950_emrld_2;Class_Probability=Alkaloid:0.410,NRP:0.880,Polyketide:0.720,RiPP:0.490,Saccharide:0.450,Terpene:0.270,Other:0.570;Edge=False\n",
      "CM000950\tEMERALDv0.1\tCLUSTER\t747262\t777103\t.\t.\t.\tID=CM000950_emrld_3;Class_Probability=Alkaloid:0.030,NRP:0.050,Polyketide:0.050,RiPP:0.030,Saccharide:0.260,Terpene:0.440,Other:0.120;Edge=False\n",
      "CM000950\tEMERALDv0.1\tCLUSTER\t917061\t974020\t.\t.\t.\tID=CM000950_emrld_4;Class_Probability=Alkaloid:0.070,NRP:0.110,Polyketide:0.200,RiPP:0.420,Saccharide:0.050,Terpene:0.020,Other:0.170;Edge=False\n",
      "CM000950\tEMERALDv0.1\tCLUSTER\t1451219\t1472608\t.\t.\t.\tID=CM000950_emrld_5;Class_Probability=Alkaloid:0.000,NRP:0.010,Polyketide:0.000,RiPP:0.400,Saccharide:0.000,Terpene:0.000,Other:0.020;Edge=False\n",
      "CM000950\tEMERALDv0.1\tCLUSTER\t1783748\t1853946\t.\t.\t.\tID=CM000950_emrld_6;Class_Probability=Alkaloid:0.190,NRP:0.510,Polyketide:0.840,RiPP:0.500,Saccharide:0.470,Terpene:0.160,Other:0.500;Edge=False\n",
      "CM000950\tEMERALDv0.1\tCLUSTER\t1894864\t1953271\t.\t.\t.\tID=CM000950_emrld_7;Class_Probability=Alkaloid:0.060,NRP:0.040,Polyketide:0.070,RiPP:0.040,Saccharide:0.410,Terpene:0.430,Other:0.160;Edge=False\n",
      "CM000950\tEMERALDv0.1\tCLUSTER\t2037759\t2065499\t.\t.\t.\tID=CM000950_emrld_8;Class_Probability=Alkaloid:0.000,NRP:0.200,Polyketide:0.060,RiPP:0.040,Saccharide:0.010,Terpene:0.000,Other:0.490;Edge=False\n",
      "CM000950\tEMERALDv0.1\tCLUSTER\t2423846\t2434683\t.\t.\t.\tID=CM000950_emrld_9;Class_Probability=Alkaloid:0.000,NRP:0.010,Polyketide:0.010,RiPP:0.740,Saccharide:0.010,Terpene:0.080,Other:0.030;Edge=False\n",
      "CM000950\tEMERALDv0.1\tCLUSTER\t2974377\t2994307\t.\t.\t.\tID=CM000950_emrld_10;Class_Probability=Alkaloid:0.110,NRP:0.020,Polyketide:0.090,RiPP:0.040,Saccharide:0.060,Terpene:0.280,Other:0.610;Edge=False\n",
      "CM000950\tEMERALDv0.1\tCLUSTER\t3638063\t3681381\t.\t.\t.\tID=CM000950_emrld_11;Class_Probability=Alkaloid:0.120,NRP:0.630,Polyketide:0.840,RiPP:0.020,Saccharide:0.300,Terpene:0.320,Other:0.640;Edge=False\n",
      "CM000950\tEMERALDv0.1\tCLUSTER\t3823745\t3855691\t.\t.\t.\tID=CM000950_emrld_12;Class_Probability=Alkaloid:0.000,NRP:0.110,Polyketide:0.050,RiPP:0.030,Saccharide:0.020,Terpene:0.000,Other:0.380;Edge=False\n",
      "CM000950\tEMERALDv0.1\tCLUSTER\t3969841\t3986931\t.\t.\t.\tID=CM000950_emrld_13;Class_Probability=Alkaloid:0.050,NRP:0.060,Polyketide:0.070,RiPP:0.650,Saccharide:0.030,Terpene:0.000,Other:0.090;Edge=False\n",
      "CM000950\tEMERALDv0.1\tCLUSTER\t4546490\t4579641\t.\t.\t.\tID=CM000950_emrld_14;Class_Probability=Alkaloid:0.090,NRP:0.250,Polyketide:0.360,RiPP:0.040,Saccharide:0.080,Terpene:0.180,Other:0.400;Edge=False\n",
      "CM000950\tEMERALDv0.1\tCLUSTER\t4754043\t4780527\t.\t.\t.\tID=CM000950_emrld_15;Class_Probability=Alkaloid:0.100,NRP:0.010,Polyketide:0.090,RiPP:0.070,Saccharide:0.070,Terpene:0.200,Other:0.360;Edge=False\n",
      "CM000950\tEMERALDv0.1\tCLUSTER\t5129720\t5146039\t.\t.\t.\tID=CM000950_emrld_16;Class_Probability=Alkaloid:0.050,NRP:0.320,Polyketide:0.080,RiPP:0.110,Saccharide:0.000,Terpene:0.000,Other:0.660;Edge=False\n",
      "CM000950\tEMERALDv0.1\tCLUSTER\t5271093\t5273041\t.\t.\t.\tID=CM000950_emrld_17;Class_Probability=Alkaloid:0.000,NRP:0.000,Polyketide:0.010,RiPP:0.165,Saccharide:0.000,Terpene:0.970,Other:0.130;Edge=False\n",
      "CM000950\tEMERALDv0.1\tCLUSTER\t5341020\t5364603\t.\t.\t.\tID=CM000950_emrld_18;Class_Probability=Alkaloid:0.030,NRP:0.020,Polyketide:0.140,RiPP:0.000,Saccharide:0.580,Terpene:0.000,Other:0.010;Edge=False\n",
      "CM000950\tEMERALDv0.1\tCLUSTER\t5464903\t5476144\t.\t.\t.\tID=CM000950_emrld_19;Class_Probability=Alkaloid:0.000,NRP:0.000,Polyketide:0.000,RiPP:0.551,Saccharide:0.000,Terpene:0.000,Other:0.000;Edge=False\n",
      "CM000950\tEMERALDv0.1\tCLUSTER\t5516963\t5527531\t.\t.\t.\tID=CM000950_emrld_20;Class_Probability=Alkaloid:0.000,NRP:0.460,Polyketide:0.730,RiPP:0.002,Saccharide:0.120,Terpene:0.240,Other:0.420;Edge=False\n",
      "CM000950\tEMERALDv0.1\tCLUSTER\t6164083\t6192894\t.\t.\t.\tID=CM000950_emrld_21;Class_Probability=Alkaloid:0.270,NRP:0.700,Polyketide:0.470,RiPP:0.010,Saccharide:0.200,Terpene:0.220,Other:0.710;Edge=False\n",
      "CM000950\tEMERALDv0.1\tCLUSTER\t6537370\t6557814\t.\t.\t.\tID=CM000950_emrld_22;Class_Probability=Alkaloid:0.060,NRP:0.010,Polyketide:0.100,RiPP:0.030,Saccharide:0.090,Terpene:0.010,Other:0.450;Edge=False\n",
      "CM000950\tEMERALDv0.1\tCLUSTER\t6792113\t6832578\t.\t.\t.\tID=CM000950_emrld_23;Class_Probability=Alkaloid:0.130,NRP:0.040,Polyketide:0.210,RiPP:0.140,Saccharide:0.060,Terpene:0.150,Other:0.370;Edge=False\n",
      "CM000950\tEMERALDv0.1\tCLUSTER\t7526991\t7555965\t.\t.\t.\tID=CM000950_emrld_24;Class_Probability=Alkaloid:0.140,NRP:0.110,Polyketide:0.090,RiPP:0.080,Saccharide:0.020,Terpene:0.320,Other:0.470;Edge=False\n",
      "CM000950\tEMERALDv0.1\tCLUSTER\t7693037\t7865528\t.\t.\t.\tID=CM000950_emrld_25;Class_Probability=Alkaloid:0.480,NRP:0.840,Polyketide:0.960,RiPP:0.140,Saccharide:0.810,Terpene:0.340,Other:0.640;Edge=False\n",
      "CM000950\tEMERALDv0.1\tCLUSTER\t7873080\t8025920\t.\t.\t.\tID=CM000950_emrld_26;Class_Probability=Alkaloid:0.460,NRP:0.920,Polyketide:0.880,RiPP:0.110,Saccharide:0.660,Terpene:0.410,Other:0.650;Edge=False\n"
     ]
    }
   ],
   "source": [
    "\n",
    "!grep \"TER\\s\" /Users/fragoso/modules/emeraldbgc/emeraldbgc/modules/CM000950.fna.prodigal.faa.emrld/CM000950.fna.prodigal.faa.emerald.full.gff2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e4bd8d4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import groupby"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ac455bc8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "218 250\n",
      "587 629\n",
      "657 684\n",
      "810 867\n",
      "1280 1297\n",
      "1553 1605\n",
      "1643 1690\n",
      "1757 1777\n",
      "2081 2086\n",
      "2552 2570\n",
      "3169 3198\n",
      "3316 3345\n",
      "3453 3469\n",
      "3943 3990\n",
      "4139 4161\n",
      "4439 4452\n",
      "4537 4538\n",
      "4594 4608\n",
      "4702 4712\n",
      "4745 4755\n",
      "5315 5340\n",
      "5655 5672\n",
      "5884 5926\n",
      "6527 6548\n",
      "6672 6791\n",
      "6800 6918\n"
     ]
    }
   ],
   "source": [
    "c = 'CM000950'\n",
    "\n",
    "for k,g in groupby(enumerate(annotate.looseClst[c]),key = lambda z:z[1]):\n",
    "    if k ==0:\n",
    "        continue\n",
    "    gg = list(zip(*g))\n",
    "    print(gg[0][0],gg[0][-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e04fca21",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 0, 0, ..., 0, 0, 0])"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "annotate.looseClst['CM000950']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73a6170d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
